lib/app/service/docker_test.go

```diff
@@ -83,3 +98,184 @@ func (s *VendorSuite) TestGeneratesProperPackageName(c *C) {
 		c.Assert(*runtimePackage, compare.DeepEquals, testCase.result, comment)
 	}
 }
+
+const imageRepository = "test/image"
+
+func generateDockerImage(dockerClient *dockerapi.Client) (*loc.DockerImage, error) {
+	testImageVersion := rand.Int()
+	image := loc.DockerImage{
+		Repository: imageRepository,
+		Tag:        strconv.Itoa(testImageVersion),
+	}
+	imageName := image.String()
+	dockerFile := fmt.Sprintf("FROM busybox\nRUN echo \"test image %s\" > test.txt", imageName)
+	testImageContextDir, err := ioutil.TempDir("", "test-image-")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer os.RemoveAll(testImageContextDir)
+	err = ioutil.WriteFile(filepath.Join(testImageContextDir, "Dockerfile"), []byte(dockerFile), 0600)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	err = dockerClient.BuildImage(dockerapi.BuildImageOptions{
+		Name:         imageName,
+		ContextDir:   testImageContextDir,
+		OutputStream: os.Stdout,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &image, nil
+}
+
+func generateDockerImages(dockerClient *dockerapi.Client, size int) ([]loc.DockerImage, error) {
+	images := make([]loc.DockerImage, 0, size)
+	for i := 0; i < size; i++ {
+		image, err := generateDockerImage(dockerClient)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		images = append(images, *image)
+	}
+	return images, nil
+}
+
+func (s *VendorSuite) TestPullAndPushImages(c *C) {
```

Comment:
I have to be honest - the test is pretty tedious to read. First of all, it's flooding with irrelevant details that blurs the actual test points. Also it tests multiple things at the same time which is not ideal but could be tolerable iff the test flow can be recognized. I would suggest to factor out bits of code that make up for all of the configuration noise and try to keep the test itself focused on minimal setup and assert points:

Assuming we've factored everything out, this would read easier:
```go
func (s *DockerSuite) TestPullAndPushImages(c *C) {
	// Setup
	const dockerImageSize = 6

	dockerImages := s.generateDockerImages(dockerImageSize, c)
	defer s.removeImages(dockerImages)

	// the first 3 docker images are pushed to both registries
	var pushedDockerTags []string
	for _, image := range dockerImages[:3] {
		s.src.push(image, c)
		s.dst.push(image, c)
		pushedDockerTags = append(pushedDockerTags, image.Tag)
	}
	sort.Strings(pushedDockerTags)

	// the last docker images are pushed only to the source registry
	var unpushedDockerTags []string
	for _, image := range dockerImages[3:] {
		s.src.push(image, c)
		unpushedDockerTags = append(unpushedDockerTags, image.Tag)
	}

	allDockerTags, exportedImages := splitAsTagsAndImages(dockerImages, s.src.r.Addr())
	srcImageRepository := fmt.Sprintf("%s/%s", s.src.r.Addr(), imageRepository)
	localImageTags := s.listTags(srcImageRepository, c)

	// generated docker images should not be in the local docker registry
	for _, tag := range allDockerTags {
		if localImageTags[tag] {
			c.Errorf("image %s:%s should not be in the local docker registry", imageRepository, tag)
		}
	}

	// all docker images should be in the source docker registry
	srcTags, err := s.helper.ImageTags(context.Background(), s.src.info.getURL(), imageRepository)
	c.Assert(err, IsNil)
	sort.Strings(srcTags)
	c.Assert(srcTags, DeepEquals, allDockerTags)

	// only pushed docker images should be in the target docker registry
	dstTags, err := s.helper.ImageTags(context.Background(), s.dst.info.getURL(), imageRepository)
	c.Assert(err, IsNil)
	sort.Strings(dstTags)
	c.Assert(dstTags, DeepEquals, pushedDockerTags)

	// Validate: this is where actual validation starts
	// export images
	err = s.helper.PullAndExportImages(context.Background(), exportedImages, s.dst.info, false, dockerImageSize)
	c.Assert(err, IsNil)

	// relist tags
	localImageTags = s.listTags(srcImageRepository, c)

	// only unpushed docker images should not be in the local docker registry
	for _, tag := range unpushedDockerTags {
		if !localImageTags[tag] {
			c.Errorf("image %s:%s should be in the local docker registry", srcImageRepository, tag)
		}
	}
	for _, tag := range pushedDockerTags {
		if localImageTags[tag] {
			c.Errorf("image %s:%s should not be in the local docker registry", srcImageRepository, tag)
		}
	}

	// all docker images should be in the target docker registry
	dstTags, err = s.helper.ImageTags(context.Background(), s.dst.info.getURL(), imageRepository)
	c.Assert(err, IsNil)
	sort.Strings(dstTags)
	c.Assert(dstTags, DeepEquals, allDockerTags)
}
```

I'd go with something like this to actually implement it:

```go
// Set up a separate Suite for this test so we can use SetUp/TearDown phases
type DockerSuite struct {
	client   *dockerapi.Client
	helper   *dockerHelper
	src, dst *registry
}

func (s *DockerSuite) SetUpTest(c *C) {
	var err error
	s.client, err = docker.NewClientFromEnv()
	c.Assert(err, IsNil)
	s.helper = newDockerHelper(logrus.New(), s.client, utils.DiscardProgress)
	// Set up source and destination registries
	s.src = s.newRegistry(c)
	s.dst = s.newRegistry(c)
}

func (s *DockerSuite) TearDownTest(*C) {
	s.src.r.Close()
	s.dst.r.Close()
}
// ...
func (s *DockerSuite) listTags(repository string, c *C) (tags map[string]bool) {
	opts := dockerapi.ListImagesOptions{Filters: map[string][]string{
		"reference": {repository},
	}}
	images, err := s.client.ListImages(opts)
	c.Assert(err, IsNil)
	tags = make(map[string]bool)
	for _, image := range images {
		for _, imageName := range image.RepoTags {
			dockerImage, err := loc.ParseDockerImage(imageName)
			c.Assert(err, IsNil)
			tags[dockerImage.Tag] = true
		}
	}
	return tags
}

// newRegistry returns a new started docker registry
func (s *DockerSuite) newRegistry(c *C) *registry {
	dir := c.MkDir()
	config := docker.BasicConfiguration("127.0.0.1:0", dir)
	r, err := docker.NewRegistry(config)
	c.Assert(err, IsNil)
	c.Assert(r.Start(), IsNil)
	return &registry{
		r:   r,
		dir: dir,
		info: registryInfo{
			address:  r.Addr(),
			protocol: "http",
		},
		helper: s.helper,
	}
}

func (s *DockerSuite) generateDockerImage(version string, c *C) loc.DockerImage {
	image := loc.DockerImage{
		Repository: imageRepository,
		Tag:        version,
	}
	imageName := image.String()
	// Use scratch for throw-away images
	dockerFile := []byte("FROM scratch")
	c.Assert(s.client.BuildImage(dockerapi.BuildImageOptions{
		Name:         imageName,
		// Use build context from an input stream to forgo the directory/file store
		InputStream:  bytes.NewReader(dockerFile),
		OutputStream: os.Stdout,
	}), IsNil)
	return image
}

func (s *DockerSuite) generateDockerImages(size int, c *C) []loc.DockerImage {
	// Use a predictable tagging scheme
	imageName := func(i int) string {
		return fmt.Sprintf("v%d.0", i)
	}
	images := make([]loc.DockerImage, 0, size)
	for i := 0; i < size; i++ {
		image := s.generateDockerImage(imageName(i), c)
		images = append(images, image)
	}
	return images
}

func (s *DockerSuite) removeImages(images []loc.DockerImage) {
	for _, image := range images {
		// Error is ignored since this is a best-effort cleanup
		s.client.RemoveImage(image.String())
	}
}

func (r *registry) push(image loc.DockerImage, c *C) {
	c.Assert(r.helper.PushImage(image.String(), r.r.Addr()), IsNil)
}

type registry struct {
	dir    string
	r      *docker.Registry
	info   registryInfo
	helper *dockerHelper
}

func splitAsTagsAndImages(images []loc.DockerImage, regAddr string) (tags, exportedImages []string) {
	for _, image := range images {
		tags = append(tags, image.Tag)

		exportedImage := image
		exportedImage.Registry = regAddr
		exportedImages = append(exportedImages, exportedImage.String())
	}
	sort.Strings(tags)
	return tags, exportedImages
}

const imageRepository = "test/image"
```